Update dependencies

davidbailey00 opened the pull request:
Update dependencies to latest versions, notably:

- `eslint` v6 to v7 - none of the [breaking changes](https://eslint.org/blog/2020/05/eslint-v7.0.0-released#breaking-changes) affect the current setup of this project
- `rimraf` v2 to v3 - [drops support for Node.js 6 and older](https://github.com/isaacs/rimraf/blob/master/CHANGELOG.md#v30)
- `style-loader` v0 to v1 - none of the [breaking changes](https://github.com/webpack-contrib/style-loader/releases/tag/v1.0.0) affect the current setup of this project
- `@jupyter-widgets/base` v2 to v3 - only [dependency updates](https://github.com/jupyter-widgets/ipywidgets/compare/@jupyter-widgets/base@2.0.2...@jupyter-widgets/base@3.0.0) as far as I can tell (considering changes in the `packages/base` directory)
- `fs-extra` v4 to v9 - [various breaking changes](https://github.com/jprichardson/node-fs-extra/blob/master/CHANGELOG.md) which are mostly bugfixes, notably Node.js v10+ is required

This seems to also fix #238 so may be worth closing #245 if this is correct - I think the newest version of `@jupyter-widgets/base` seems to inject stylesheets in the correct order consistently

martinRenou commented:
Thanks for your PR!

Could you please revert the change on the `@jupyter-widgets/base` dependency though? We use this to allow support for JupyterLab 1 and 2. If we keep your change, only JupyterLab 2 will be supported.

ianhi commented on js/package.json:
```diff
@@ -27,18 +27,18 @@
+    "webpack-cli": "^3.3.12"
   },
   "dependencies": {
-    "@jupyter-widgets/base": "^2 || ^3"
```
I think that the `^2 || ^3` is what allows simultaneous support of jupyterlab 1 and 2.

ianhi replied:
oh wow. i sure didn't see the other comment already saying as much before i wrote that...

davidbailey00 commented:
Hi, sorry for leaving this for so long - I have just updated to the latest versions again and fixed merge conflicts, since this PR had since become out-of-date with the main branch :)

I just want to clarify on this:

> Could you please revert the change on the @jupyter-widgets/base dependency though? We use this to allow support for JupyterLab 1 and 2. If we keep your change, only JupyterLab 2 will be supported.

The way NPM packages work means that for `dependencies` (and `devDependencies`), the latest version which fits a range will always be used when the package is installed by a user. So for example, `^2 || ^3 || ^4.0.0` will currently resolve to `4.0.0` regardless of the version of JupyterLab the user has. Therefore, updating the range to `^4` will still work the same given that the latest version is in this range.

If this is not what you expected, it may be that `@jupyter-widgets/base` should be a `peerDependency` instead of a `dependency`. Peer dependencies are different in that the consumer of a dependency (in this case, JupyterLab) is expected to install those dependencies, rather than them being explicitly required by the package in question (i.e. this plugin).

However, if the plugin currently works as expected in JupyterLab v1 then I would advise keeping this as a `dependency` and bumping the range to only allow version 4 should make no difference :)

martinRenou commented:
Everything you said is true for the nbextension, the one that we build in this package and that we ship to npmjs. 

Although this is not true anymore when it comes to JupyterLab. JupyterLab has its own dependency solver (using yarn) that it uses to build itself in v1 and v2. If we enforce `@jupyter-widgets/base` v4, JupyterLab v1 and v2 will not build anymore with the ipympl labextension, because it will find a dependency conflict between `@jupyter-widgets/base`, `@jupyter-widgets/jupyterlab-manager`, `ipympl` and JupyterLab itself.

jasongrout commented:
In particular, JupyterLab uses yarn resolutions to force specific versions of the core packages. `@jupyter-widgets/base` depends on different versions of the core packages as well, and the major versions of `@jupyter-widgets/base` are tied to specific JupyterLab major versions. This, combined with a deduplication step that happens in a JupyterLab build, forces the JupyterLab build to choose specific versions of `@jupyter-widgets/base` for each major version of JupyterLab. That is why the version range for dependencies needs to be wide enough to allow older versions of `@jupyter-widgets/base` for older versions of JupyterLab.

jasongrout commented:
> If this is not what you expected, it may be that `@jupyter-widgets/base` should be a `peerDependency` instead of a `dependency`. Peer dependencies are different in that the consumer of an NPM package (the consumer being JupyterLab in this case) is expected to install those dependencies, rather than those being explicitly required by the dependency (i.e. this plugin).

Hmm, I wonder if we might be able to use peer dependencies. It seems like we should have considered this at some point, but I can't recall any specific reason we could not use peer dependencies (it's also late-I might be forgetting something). I'll think about it (or if someone else has a good answer for why peer dependencies won't work, please comment!)

martinRenou commented:
> It seems like we should have considered this at some point, but I can't recall any specific reason we could not use peer dependencies

This might work for the JupyterLab build to have `jupyter-widgets/base` as a peer dependency, but not for the nbextension build as it would be missing?

jasongrout commented:
It would need it as a dev dependency and a peer dependency, similar to how I think a lot of react packages work?

martinRenou commented:
Oh ok :) I didn't know about this kind of setup.

jasongrout commented:
I'd recommend reverting to what we know works here, and let us think through peer dependencies over in jlab and come up with a recommendation.

davidbailey00 commented:
Thank you both for the explanations, apologies for my wrong assumption about how this works - I will revert back to using the more inclusive range

davidbailey00 commented:
Okay, reverted to the range!

martinRenou commented:
Closing as outdated
